tests/unittests/unittests.cpp

```diff
@@ -538,6 +557,58 @@ class ReaderWriterQueueTests : public TestClass<ReaderWriterQueueTests>
 
 		return true;
 	}
+
+#if MOODYCAMEL_HAS_EMPLACE
+	bool emplace()
+	{
+		ReaderWriterQueue<UniquePtrWrapper> q(100);
+		std::unique_ptr<int> p { new int(123) };
+		q.emplace(std::move(p));
+		UniquePtrWrapper item;
+		ASSERT_OR_FAIL(q.try_dequeue(item));
+		ASSERT_OR_FAIL(item.get_value() == 123);
+		ASSERT_OR_FAIL(q.size_approx() == 0);
+
+		return true;
+	}
+
+	// This is what you have to do to try_enqueue() a movable type, and demonstrates why try_emplace() is useful
+	bool try_enqueue_bad()
+	{
+		ReaderWriterQueue<UniquePtrWrapper> q(0);
+		{
+			// A failed try_enqueue() will still delete p
+			std::unique_ptr<int> p { new int(123) };
+			q.try_enqueue(std::move(p));
+			ASSERT_OR_FAIL(q.size_approx() == 0);
+			ASSERT_OR_FAIL(p == nullptr);
+		}
+		{
+			// Workaround isn't pretty and potentially expensive - use try_emplace() instead
+			std::unique_ptr<int> p { new int(123) };
+			UniquePtrWrapper w(std::move(p));
+			q.try_enqueue(std::move(w));
+			p = std::move(w.get_ptr());
+			ASSERT_OR_FAIL(q.size_approx() == 0);
+			ASSERT_OR_FAIL(p != nullptr);
+			ASSERT_OR_FAIL(*p == 123);
+		}
+
+		return true;
+	}
+
+	bool try_emplace()
```

Comment:
Hmm, maybe rename this to `try_emplace_fail`?

Reply:
Done.